contributors/design-proposals/lifecycle-hooks.md

```diff
@@ -0,0 +1,152 @@
+It is mostly used to run some kind of acceptance check in the middle and/or at the end to fully verify the update is working as expected and rollback if it isn't. 
+The acceptance check may be time consuming and more thorough than what readiness and liveness probes are intended for. 
+You can also notify external services from them, migrate database in the middle of an update, send messages to IRC channel or do anything else.
+
+[A short demo](https://youtu.be/GVNTm_K43iI) simulating lifecycle hooks using auto-pausing that has been presented at SIG-Apps meeting on August 21, 2017.
+
+
+## API Objects
+### New Objects
+```go
+type HookTriggerType string
+
+const (
+	ProgressAvailableHookTriggerType HookTriggerType = "ProgressAvailable"
+)
+
+type ProgressAvailableParams struct {
+    // ProgressPoint specifies the point during update when the hook should be triggered.
+    // Accepts both the number of new pods available or a percentage value representing the ratio
+    // of new.availableReplicas to the declared replicas. In case of getting a percentage
+    // it will try to reach the exact or the closest possible point right after it,
+    // trigger the job, wait for it to complete and then continue. 
+    // If such situation shall occur that two different ProgressPoints should be reached at
+    // the same time, all the hooks for an earlier ProgressPoint will be ran (and finished)
+    // before any later one.
+    ProgressPoint intstr.IntOrString
+}
+
+type HookTrigger struct {
+    // Type of hook trigger. Can be ProgressAvailable. Default is ProgressAvailable.
+    Type HookTriggerType
+    
+    // ProgressAvailable config params. Present only if type = "ProgressAvailable".
+    ProgressAvailableParams *ProgressAvailableParams
+}
+
+type LifecycleHook struct {
+    // Name of the hook
+    Name string
+    
+    // Triggering config
+    Trigger HookTrigger
+    
+    // "Abort" - Failure for this hook is fatal, deployment becomes perma-failed
+    // "Ignore" - Even if the hook fails deployment continues rolling out new version.
+    FailurePolicy string
+    
+    // "Always" - always keep this jobs 
+    // "OnFailure" - keep this Job only if it fails
+    // "Never" - delete this job immediately after it is done 
+    RetainPolicy string
+    
+    JobSpec v1.JobSpec
+}
+```
+
+### Affected Objects
+As of now the lifecycle hooks are relevant for Deployments, StatefulSets and DaemonSets.
+New optional field named `LifecycleHooks` will be added to provide configuration for lifecycle hooks. 
+The related state will be reported using Conditions and details will be available on the created Job objects.
+Also the controllers will need to be adjusted to scale in appropriate chunks and trigger the hooks.
+
+```go
+type DeploymentSpec struct {
+    // ...
+    // Addition
+    // List of lifecycle hooks
+    // Can have multiple hooks at the same ProgressPoint; order independent
+    LifecycleHooks []LifecycleHook
```

Comment:
What are the merging semantics for these lists?

cc @apelisse

Reply:
`LifecycleHook.Name` is the unique key, when merging 2 list the newer `LifecycleHook` replaces the old one where the key matches.

Not sure if there is a syntax to express it or hint apimachinery.

Reply:
Here is an [example](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2075-L2077).

PS: We are in the process of changing the name and semantic of these "hints", but for now you should be using `patchMergeKey` and `patchStrategy`.